src/odb_backend.c

```diff
@@ -36,6 +36,182 @@
+    if (!PyArg_ParseTuple(result, "Ony#",
+                &py_oid_out, type, &bytes, sz) || !bytes)
+        return GIT_EUSER;
+
+    /* XXX: This assumes the default libgit2 allocator is in use and will
+     * probably segfault and/or destroy the universe otherwise */
+    *ptr = malloc(*sz);
+    if (!*ptr)
+        return GIT_EUSER;
+
+    memcpy(*ptr, bytes, *sz);
+    py_oid_to_git_oid(py_oid_out, oid_out);
+    return 0;
+}
+
+static int
+pygit2_odb_backend_read_header(size_t *len, git_object_t *type,
+        git_odb_backend *_be, const git_oid *oid)
+{
+    int err;
+    PyObject *args, *py_oid, *result;
+    struct pygit2_odb_backend *be = (struct pygit2_odb_backend *)_be;
+
+    py_oid = git_oid_to_python(oid);
+    args = Py_BuildValue("(O)", py_oid);
+    result = PyObject_CallObject(be->read_header_callable, args);
+    Py_DECREF(py_oid);
+
+    if ((err = git_error_for_exc()) != 0) {
+        return err;
+    }
+
+    if (result == NULL)
+        return GIT_EUSER;
+
+    if (!PyArg_ParseTuple(result, "nn", type, len))
+        return GIT_EUSER;
+
+    return 0;
+}
+
+int
+OdbBackend_init(OdbBackend *self, PyObject *args, PyObject *kwds)
+{
+    if (args && PyTuple_Size(args) > 0) {
+        PyErr_SetString(PyExc_TypeError,
+                        "OdbBackend takes no arguments");
+        return -1;
+    }
+
+    if (kwds && PyDict_Size(kwds) > 0) {
+        PyErr_SetString(PyExc_TypeError,
+                        "OdbBackend takes no keyword arguments");
+        return -1;
+    }
+
+    struct pygit2_odb_backend *be = calloc(1, sizeof(struct pygit2_odb_backend));
+    be->backend.version = GIT_ODB_BACKEND_VERSION;
+    be->OdbBackend = (PyObject *)self;
+
+    be->read_callable = PyObject_GetAttrString((PyObject *)self, "read");
+    if (be->read_callable) {
+        be->backend.read = pygit2_odb_backend_read;
+        Py_INCREF(be->read_callable);
+    }
+
+    be->read_prefix_callable = PyObject_GetAttrString(
+            (PyObject *)self, "read_prefix");
+    if (be->read_prefix_callable) {
+        be->backend.read_prefix = pygit2_odb_backend_read_prefix;
+        Py_INCREF(be->read_prefix_callable);
+    }
+
+    be->read_header_callable = PyObject_GetAttrString(
+            (PyObject *)self, "read_header");
+    if (be->read_header_callable) {
+        be->backend.read_header = pygit2_odb_backend_read_header;
+        Py_INCREF(be->read_header_callable);
+    }
+
+    self->odb_backend = (git_odb_backend *)be;
+    return 0;
+}
+
+void
+OdbBackend_dealloc(OdbBackend *self)
+{
+    if (self->odb_backend && self->odb_backend->read == pygit2_odb_backend_read) {
+        struct pygit2_odb_backend *be = (struct pygit2_odb_backend *)self->odb_backend;
+
+        if (be->read_callable)
+            Py_DECREF(be->read_callable);
+        if (be->read_prefix_callable)
+            Py_DECREF(be->read_prefix_callable);
+        if (be->read_header_callable)
+            Py_DECREF(be->read_header_callable);
```

Comment:
You can simply call `Py_CLEAR` to save the `if`, https://docs.python.org/3/c-api/refcounting.html#c.Py_CLEAR